## Differece in r2_score of  ANN Model, if we use :

* StandardScaler on our data.

* MinMaxScaler on our data. 

* Not use any scalling on our data.

In [1]:
import pandas as pd
import numpy as np

In [26]:
from sklearn.datasets import fetch_california_housing

In [27]:
data = fetch_california_housing(as_frame=True)

In [28]:
df = data.data
df[data.target_names[0]] = data.target

In [29]:
df

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0      8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1      8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2      7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3      5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4      3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   
...       ...       ...       ...        ...         ...       ...       ...   
20635  1.5603      25.0  5.045455   1.133333       845.0  2.560606     39.48   
20636  2.5568      18.0  6.114035   1.315789       356.0  3.122807     39.49   
20637  1.7000      17.0  5.205543   1.120092      1007.0  2.325635     39.43   
20638  1.8672      18.0  5.329513   1.171920       741.0  2.123209     39.43   
20639  2.3886      16.0  5.254717   1.162264      1387.0  2.616981     39.37   

       Longitude  MedHouseVal  
0        -122.23        4.526  
1        -122.22        3.585  
2        -122.24        3.521  
3        -122.25        3.413  
4        -122.25        3.422  
...          ...          ...  
20635    -121.09        0.781  
20636    -121.21        0.771  
20637    -121.22        0.923  
20638    -121.32        0.847  
20639    -121.24        0.894  

[20640 rows x 9 columns]

In [30]:
x = df.drop(columns='MedHouseVal')
y = df['MedHouseVal']

## ANN with StandardScaler

In [40]:
from sklearn.model_selection import train_test_split

In [122]:
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.2, random_state=42)

In [123]:
from sklearn.preprocessing import StandardScaler

In [124]:
ss = StandardScaler()
x_train = ss.fit_transform(x_train)
x_test = ss.transform(x_test)

In [125]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.layers import LeakyReLU, ReLU, ELU, PReLU

In [126]:
regression = Sequential()
regression.add(Dense(10, input_shape=(8,),activation='relu'))
regression.add(Dense(10,activation='relu'))
regression.add(Dense(10, activation='relu'))
regression.add(Dense(1))

In [127]:
regression.summary()

Model: "sequential_6"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_20 (Dense)            (None, 10)                90        
                                                                 
 dense_21 (Dense)            (None, 10)                110       
                                                                 
 dense_22 (Dense)            (None, 10)                110       
                                                                 
 dense_23 (Dense)            (None, 1)                 11        
                                                                 
Total params: 321
Trainable params: 321
Non-trainable params: 0
_________________________________________________________________


In [128]:
regression.compile(optimizer='adam',loss="mse",metrics=['mse'])

In [129]:
import tensorflow as tf
early_stopping = tf.keras.callbacks.EarlyStopping(
    monitor="mse",
    patience=5)

In [130]:
regression.fit(x_train, y_train, validation_data=(x_test, y_test), batch_size=10,epochs=100,callbacks=early_stopping)

Epoch 1/100
1652/1652 [==============================] - 5s 3ms/step - loss: 0.7498 - mse: 0.7498 - val_loss: 0.4260 - val_mse: 0.4260
Epoch 2/100
1652/1652 [==============================] - 5s 3ms/step - loss: 0.3962 - mse: 0.3962 - val_loss: 0.3878 - val_mse: 0.3878
Epoch 3/100
1652/1652 [==============================] - 5s 3ms/step - loss: 0.3720 - mse: 0.3720 - val_loss: 0.3711 - val_mse: 0.3711
Epoch 4/100
1652/1652 [==============================] - 5s 3ms/step - loss: 0.3544 - mse: 0.3544 - val_loss: 0.3521 - val_mse: 0.3521
Epoch 5/100
1652/1652 [==============================] - 5s 3ms/step - loss: 0.3443 - mse: 0.3443 - val_loss: 0.3527 - val_mse: 0.3527
Epoch 6/100
1652/1652 [==============================] - 5s 3ms/step - loss: 0.3365 - mse: 0.3365 - val_loss: 0.3384 - val_mse: 0.3384
Epoch 7/100
1652/1652 [==============================] - 5s 3ms/step - loss: 0.3310 - mse: 0.3310 - val_loss: 0.3407 - val_mse: 0.3407
Epoch 8/100
1652/1652 [==============================] 

In [133]:
y_preds = regression.predict(x_test)
y_preds

array([[0.573671 ],
       [1.5140991],
       [4.8436522],
       ...,
       [4.8839073],
       [0.7137127],
       [2.1245642]], dtype=float32)

In [134]:
from sklearn.metrics import r2_score

In [135]:
r2_score(y_test, y_preds)

0.7854370112022594

## ANN without  StandardScaler

In [105]:
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.2, random_state=42)

In [106]:
regression = Sequential()
regression.add(Dense(10, input_shape=(8,),activation='relu'))
regression.add(Dense(10,activation='relu'))
regression.add(Dense(10, activation='relu'))
regression.add(Dense(1))

In [107]:
regression.compile(optimizer='adam',loss="mse",metrics=['mse'])

In [108]:
import tensorflow as tf
early_stopping = tf.keras.callbacks.EarlyStopping(
    monitor="mse",
    patience=5)

In [109]:
regression.fit(x_train, y_train, validation_data=(x_test, y_test), batch_size=10,epochs=100,callbacks=early_stopping)

Epoch 1/100
1652/1652 [==============================] - 6s 3ms/step - loss: 524.9539 - mse: 524.9539 - val_loss: 1.7461 - val_mse: 1.7461
Epoch 2/100
1652/1652 [==============================] - 5s 3ms/step - loss: 1.6014 - mse: 1.6014 - val_loss: 1.1163 - val_mse: 1.1163
Epoch 3/100
1652/1652 [==============================] - 5s 3ms/step - loss: 1.7816 - mse: 1.7816 - val_loss: 0.9042 - val_mse: 0.9042
Epoch 4/100
1652/1652 [==============================] - 5s 3ms/step - loss: 2.0795 - mse: 2.0795 - val_loss: 18.0646 - val_mse: 18.0646
Epoch 5/100
1652/1652 [==============================] - 5s 3ms/step - loss: 5.0445 - mse: 5.0445 - val_loss: 0.7462 - val_mse: 0.7462
Epoch 6/100
1652/1652 [==============================] - 5s 3ms/step - loss: 8.8156 - mse: 8.8156 - val_loss: 0.8424 - val_mse: 0.8424
Epoch 7/100
1652/1652 [==============================] - 5s 3ms/step - loss: 3.5277 - mse: 3.5277 - val_loss: 0.7841 - val_mse: 0.7841


In [110]:
y_preds = regression.predict(x_test)

In [111]:
r2_score(y_test, y_preds)

0.4016325779564255

## ANN with MinMaxScaler

In [112]:
from sklearn.preprocessing import MinMaxScaler

In [113]:
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.2, random_state=42)

In [114]:
mm = MinMaxScaler()

In [115]:
x_train = mm.fit_transform(x_train)
x_test = mm.transform(x_test)

In [116]:
regression = Sequential()
regression.add(Dense(10, input_shape=(8,),activation='relu'))
regression.add(Dense(10,activation='relu'))
regression.add(Dense(10, activation='relu'))

regression.add(Dense(1))

In [117]:
regression.compile(optimizer='adam',loss="mse",metrics=['mse'])

In [118]:
import tensorflow as tf
early_stopping = tf.keras.callbacks.EarlyStopping(
    monitor="mse",
    patience=5)

In [119]:
regression.fit(x_train, y_train, validation_data=(x_test, y_test), batch_size=10,epochs=100,callbacks=early_stopping)

Epoch 1/100
1652/1652 [==============================] - 5s 3ms/step - loss: 1.0190 - mse: 1.0190 - val_loss: 0.6033 - val_mse: 0.6033
Epoch 2/100
1652/1652 [==============================] - 5s 3ms/step - loss: 0.5747 - mse: 0.5747 - val_loss: 0.5662 - val_mse: 0.5662
Epoch 3/100
1652/1652 [==============================] - 5s 3ms/step - loss: 0.5421 - mse: 0.5421 - val_loss: 0.5463 - val_mse: 0.5463
Epoch 4/100
1652/1652 [==============================] - 5s 3ms/step - loss: 0.5114 - mse: 0.5114 - val_loss: 0.5088 - val_mse: 0.5088
Epoch 5/100
1652/1652 [==============================] - 5s 3ms/step - loss: 0.4864 - mse: 0.4864 - val_loss: 0.4908 - val_mse: 0.4908
Epoch 6/100
1652/1652 [==============================] - 5s 3ms/step - loss: 0.4710 - mse: 0.4710 - val_loss: 0.4865 - val_mse: 0.4865
Epoch 7/100
1652/1652 [==============================] - 5s 3ms/step - loss: 0.4636 - mse: 0.4636 - val_loss: 0.4768 - val_mse: 0.4768
Epoch 8/100
1652/1652 [==============================] 

In [120]:
y_preds = regression.predict(x_test)

In [121]:
r2_score(y_test, y_preds)

0.735306701511156

In [136]:
0.7854370112022594 - 0.735306701511156

0.050130309691103436

In [139]:
0.735306701511156*100

73.5306701511156

In [140]:
0.7854370112022594*100

78.54370112022593

In [142]:
0.4016325779564255*100

40.16325779564255

In [141]:
78.54370112022593 - 73.5306701511156

5.013030969110332

In [143]:
73.5306701511156 - 40.16325779564255

33.36741235547305

In [146]:
78.54370112022593 - 40.16325779564255

38.38044332458338

## conclusion

* r2_score of StandardScaler is `5.01% >` then r2_score of MinMaxScaler

* r2_score of MinMaxScaler is `33.3% >` then r2_score of without any scalling

* r2_score of StandardScaler is `38.3% >` then r2_score of without any scalling